Menus: allow easy reverting of changes

georgeh opened the pull request:
This PR will address #22 - Menus: allow easy reverting of changes

The goal is to provide a way to revert changes to the menu without having to trigger a full page reload. Right now the only way to cancel your changes is to navigate away and back, which is not intuitive.

The implementation will be to create a `MenuRevertButton` component and attach it to the `Menu` component by the Save and Delete buttons. When clicked it will call `MenuData.discard()` which reloads directly from the server.

![screenshot 2016-03-27 21 08 28](https://cloud.githubusercontent.com/assets/128240/14068962/397b088e-f460-11e5-8126-8a2efa291e67.png)

**Testing**

* Go into Menus
* Confirm the Cancel button is disabled
* Make some changes to a menu
* Click the Cancel button
* Confirm the changes have been reverted

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+ */
+var React = require( 'react' ),
+	classNames = require( 'classnames' ),
+	debug = require( 'debug' )( 'calypso:menus:revert-button' ); // eslint-disable-line no-unused-vars
```
We can use es6 style imports:
```javascript
import React from 'react';
import classNames from 'classnames';
import debugFactory from 'debug';

const debug = debugFactory( 'calypso:menus:revert-button' );
```

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+	classNames = require( 'classnames' ),
+	debug = require( 'debug' )( 'calypso:menus:revert-button' ); // eslint-disable-line no-unused-vars
+
+var MenuRevertButton = React.createClass( {
```
We can use `const` here :)

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+	}
+} );
+
+module.exports = MenuRevertButton;
```
es6 style would be:
```
default export MenuRevertButton;
```
Also see: http://exploringjs.com/es6/ch_modules.html

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+	debug = require( 'debug' )( 'calypso:menus:revert-button' ); // eslint-disable-line no-unused-vars
+
+var MenuRevertButton = React.createClass( {
+	componentWillMount: function() {
```
For any methods in a class, we can also write it as:
```
componentWillMount() {
```
vs
```
componentWillMount: function() {
```

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+	},
+
+	render: function() {
+		var hasChanged = this.props.menuData.get().hasChanged,
```
`const` here too.

gwwar commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,53 @@
+			} );
+
+		return (
+			<button className={ classes }
```
Perhaps consider using our button component, available in `client/components/button`.

Also see: https://wpcalypso.wordpress.com/devdocs/design/buttons for how they work in action

gwwar commented:
:+1: works as advertised. I would also suggest removing the primary class from the button since this is not a primary action.

cc @lancewillett @mcsf @folletto

georgeh commented:
@gwwar I updated the class to use the ES6 features you pointed out, and also switched out the DOM <button> for a Button component without the primary class. I updated the screenshot at the top to reflect the new style.

folletto commented:
Hello!

I agree on the need of something like this, however, we avoided putting that button there because it impairs the use of that part of the UI on mobile – reducing too much the space for the title. See what happens:

![screen shot 2016-03-28 at 12 16 44](https://cloud.githubusercontent.com/assets/4389/14077016/f913d990-f4de-11e5-8c6d-0f1142e9254a.png)

This problem is even more evident on languages that have longer words for "Save" and "Cancel".

I think we can still add the button if we want, but we need to address the mobile breakpoint. Without changing the design (which is needed, but way beyond the scope here) I see only one potential solution:

1. At small viewports, we stop showing the title and we just show the pencil icon
2. Once the icon gets clicked, the whole line gets substituted by the text field, one can type in, and once saved the line goes back to just icon + buttons.

Does it make sense? :)

georgeh commented:
@folletto Yup, that makes sense. An alternative might be to change "Cancel" to the Cross icon - that seems like it would a) keep the edit icon on the same line and b) be consistent with the delete icon.

I'll play with both solutions and get some screenshots to consider.

folletto commented:
> Edit: I ruled out the Cross icon because it didn't make any sense unless you already know what it does. Not good UI.

I just opened the thread to say that and then I saw the edit :+1:  


> I'm running into a sticking point with the pencil icon wrapping. 

Yeah I recall that being an annoying bit to work with. :|

georgeh commented:
@folletto I fixed the mobile wrapping issue, but please compare cbe6114 with Automattic/calypso-pre-oss@0ef153c - I believe the `max-width: calc(...)` + `overflow: ellipsis issue` is no longer happening but I am having trouble confirming. Do you remember what platform that was happening on?

I tried a few different configurations and took screenshots at a 320px width.

Option 1: Here's what it looks like with just the line-break fixed:
![Words and Icon](https://cloud.githubusercontent.com/assets/128240/14128754/cdf96252-f5f1-11e5-8324-3754c64c817e.png)

Option 2: Here's what it looks like if I remove the title:
![Icon only](https://cloud.githubusercontent.com/assets/128240/14128761/dad0c4a2-f5f1-11e5-808b-ee0203061b45.png)

Option 3: I also tried hiding the pencil icon instead of the title with small view ports but it hides the fact that it's editable:
![Words only](https://cloud.githubusercontent.com/assets/128240/14128767/e73fefa6-f5f1-11e5-8a48-76db1739137a.png)

Also, here's what it looks like when you get into edit mode:
![Editing](https://cloud.githubusercontent.com/assets/128240/14128780/fd644f20-f5f1-11e5-9537-0a9ea596bf97.png)

My vote is for option 1, but my read of your solution is option 2. Do you have a preference? I'll wait to commit my changes until I hear from you.

mcsf commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,54 @@
+	},
+
+	render() {
+		const hasChanged = this.props.menuData.get().hasChanged;
```
Not important, but this could use ES6 destructuring to make it slightly shorter and more concise.

georgeh replied:
The only thing I could come up with is `const { hasChanged: hasChanged } = this.props.menuData.get();` which is slightly longer and I'm not sure it adds any clarity. Am I missing a cool ES6 feature?

nb replied:
In ES2015 you can substitute `{something: something}` with just `{something}`:
```js
const { hasChanged } = this.props.menuData.get();
```

georgeh replied:
TIL! Thanks!

mcsf commented:
I haven't QA'd, but the changes look good code-wise. They also serve as a painful reminder of how horrible the Menus architecture is. :)

folletto commented:
> My vote is for option 1, but my read of your solution is option 2. Do you have a preference? I'll wait to commit my changes until I hear from you.

I understand why option 1, but gets very tight. To be honest I don't thing we have a real winner here, just varied sub-optimal solutions. For the sake of this PR, and waiting a more proper approach, let's go for Option 1.

Would be nice if possible to make the text are full width in that case, like this:

![screen shot 2016-03-31 at 12 38 23](https://cloud.githubusercontent.com/assets/4389/14174622/7aee3aea-f73d-11e5-89f4-20097d8613a2.png)

But I would consider this a nice extra and not a blocker to get this PR merged. :)

georgeh commented:
I pushed up changes to lock in Option 1 so hopefully this is good to go. I looked a bit at hiding the other buttons but I didn't want that to block this PR.

nb commented on client/my-sites/menus/menu.jsx:
```diff
@@ -242,6 +243,7 @@ var Menu = React.createClass( {
 								selectedLocation={ this.props.selectedLocation }
 								setBusy={ this.props.setBusy }
 								confirmDiscard={ this.props.confirmDiscard } />
+						<MenuRevertButton menuData={ this.props.siteMenus } />
```
Why do we call this a revert button if it says “Cancel” in the UI?

georgeh replied:
I thought about that and chose "revert" intentionally. When looking at the UI the word "cancel" is clearer about what it does for the user, but I think that "revert" explains more about what the button does at a code level. I decided that it was better for the code to describe what the code does and the UI to describe what the user wants to accomplish, than to have a generalized term for the code that isn't as clear.

nb commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,54 @@
+		return (
+			<Button disabled={ ! hasChanged }
+					onClick={ this.revert }>
+				{ this.state.reverting ? this.translate( 'Canceling…' ) : this.translate( 'Cancel' ) }
```
It’s great that you used the unicode ellipsis character here, instead of three dots :) 💌

nb commented on assets/stylesheets/sections/_menus.scss:
```diff
@@ -194,7 +194,7 @@
 			display: block;
 			float: left;
-			max-width: 90%; /* should be less to avoid a bug on touch + <a> tag that auto-closes the area, but ellipsis wouldn't work */
+			max-width: calc(100% - 20px);
```
We might want to add spaces around those parens.

nb replied:
Hm, I see we’re not terribly consistent with those spaces. Your choice.

georgeh replied:
That was bugging me so I added them.

nb commented on client/my-sites/menus/menus-revert-button.jsx:
```diff
@@ -0,0 +1,54 @@
+		return { reverting: false };
+	},
+
+	setReverting() {
```
How do you think names like `startReverting()` and `endReverting()` would semantically compare to their `set*` equivalents?

georgeh replied:
Good question, I hadn't considered those names. `reverting` and `reverted` are states, which is what these methods set. `start` and `end` seem more like events, which would imply that these methods would emit something. Also, their usage in the code seems more natural with `setReverting()` and `setReverted()` than start/end. Based on that reasoning I prefer their current names, but am interested in hearing arguments for start/end.

gwwar replied:
`setReverting` reads like the method should be passed a boolean value, so something like `startReverting()` and `endReverting()` might work better here.

georgeh replied:
Sold! That totally makes sense, I'll make the updates.

nb commented:
@georgeh, my comments are not blockers to merging, but could you rebase/squash (if needed) before we merge this?

Also, what do you think of @folletto’s suggestion to make editing the title full-width?

georgeh commented:
@nb I made some style changes (parens in SCSS and ES6 destructured assignment) and squashed/rebased.

@folletto - I played around with making the title 100% width at small widths but don't like how the buttons wrap:

![wrapped title view](https://cloud.githubusercontent.com/assets/128240/14338022/3712459c-fc41-11e5-94e5-cb6312f0f756.png)

![wrapped title edit](https://cloud.githubusercontent.com/assets/128240/14338026/436a2044-fc41-11e5-9c17-739d550078d0.png)

My vote is against wrapping to full width. I can't really justify it, it's my personal taste.

lancewillett commented:
Sorry for the bot noise... my fault.

folletto commented:
> I played around with making the title 100% width at small widths but don't like how the buttons wrap:

I agree, I don't think it should wrap. 

Editing the field should cover the entire line, see [the mockup above](https://cloud.githubusercontent.com/assets/4389/14174622/7aee3aea-f73d-11e5-89f4-20097d8613a2.png). :)

folletto commented:
Tested on Chrome, Firefox, Safari: :+1:

georgeh commented:
I pulled this back  to see if there was a cleaner way to have `MenuName` communicate its state to `Menu`, either with an event emitter or with Reflux. Neither seemed to be used very much in this component so I decided to leave the callback prop to pass the state from `MenuName` to `Menu` rather that introduce new architecture.

gwwar commented:
Unless you intend to persist this menu state through multiple page views, what you have here is appropriate. The Redux approach might involve adding an additional subtree to ui. See also: https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#ui-state

gwwar commented:
Let's disable the cancel button when we click save. We can get run into some unexpected behavior when we do the following:
1. Make some changes
2. Click Save
3. Click Cancel
4. Full page refresh

gwwar commented:
👍  Tested on Chrome, FF, IE. @georgeh ping me when that last corner case is cleared up, and I can help you deploy.

georgeh commented:
Thanks for the improvements @gwwar - I've pushed up changes that will disable the Cancel button while a save is in progress.

gwwar commented:
![hifive](https://cloud.githubusercontent.com/assets/1270189/14906893/ada0ec82-0d75-11e6-84db-6307179fbebe.gif)

Fixes #22
